File: scripts/scrape_tripadvisor_attractions.py
from __future__ import annotations

import re
import shutil
from pathlib import Path

from openpyxl import Workbook
from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
from playwright.sync_api import sync_playwright


TARGET_URL = "https://www.tripadvisor.cn/Attractions-g294211-Activities-China.html"
MAX_PAGES = 10
OUTPUT_XLSX = Path("attractionsForAgent.xlsx")
OUTPUT_VIDEO = Path("tripadvisor-attractions-capture.webm")
VIDEO_DIR = Path("video_tmp")

RANKED_NAME_PATTERN = re.compile(r"^(\d+)\.(.+)$")


def extract_ranked_attractions(page) -> list[tuple[str, str]]:
    """Extract (rank, name) from attraction cards on the current list page."""
    items: list[tuple[str, str]] = []
    cards = page.locator("article")
    for idx in range(cards.count()):
        lines = [line.strip() for line in cards.nth(idx).inner_text().splitlines() if line.strip()]
        for line in lines:
            match = RANKED_NAME_PATTERN.match(line)
            if match:
                rank = match.group(1)
                name = match.group(2).strip()
                items.append((rank, name))
                break
    return items


def save_to_excel(rows: list[tuple[int, str, str]], output_path: Path) -> None:
    workbook = Workbook()
    sheet = workbook.active
    sheet.title = "Attractions"
    sheet.append(["页码", "编号", "景点名称"])
    for row in rows:
        sheet.append(list(row))
    workbook.save(output_path)


def main() -> None:
    VIDEO_DIR.mkdir(parents=True, exist_ok=True)
    if OUTPUT_VIDEO.exists():
        OUTPUT_VIDEO.unlink()

    all_rows: list[tuple[int, str, str]] = []

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1366, "height": 768},
            locale="zh-CN",
            record_video_dir=str(VIDEO_DIR),
            record_video_size={"width": 1366, "height": 768},
        )
        page = context.new_page()
        page.goto(TARGET_URL, wait_until="domcontentloaded", timeout=120_000)
        page.wait_for_timeout(4_000)

        captured_pages = 0
        video_path = page.video.path()

        while captured_pages < MAX_PAGES:
            current_page_index = captured_pages + 1
            page_items = extract_ranked_attractions(page)
            for rank, name in page_items:
                all_rows.append((current_page_index, rank, name))

            captured_pages += 1
            if captured_pages >= MAX_PAGES:
                break

            next_button = page.locator("a[aria-label='Next page']").first
            if next_button.count() == 0:
                break

            if next_button.get_attribute("aria-disabled") == "true":
                break

            first_item_text = page_items[0][0] + "." + page_items[0][1] if page_items else ""
            next_button.click()

            if first_item_text:
                try:
                    page.wait_for_function(
                        """(needle) => {
                            const firstArticle = document.querySelector("article");
                            if (!firstArticle) return false;
                            return !firstArticle.innerText.includes(needle);
                        }""",
                        arg=first_item_text,
                        timeout=30_000,
                    )
                except PlaywrightTimeoutError:
                    page.wait_for_timeout(3_000)
            else:
                page.wait_for_timeout(3_000)

            page.wait_for_load_state("domcontentloaded")
            page.wait_for_timeout(2_000)

        save_to_excel(all_rows, OUTPUT_XLSX)

        context.close()
        browser.close()

    video_file = Path(video_path)
    if video_file.exists():
        shutil.move(str(video_file), OUTPUT_VIDEO)

    shutil.rmtree(VIDEO_DIR, ignore_errors=True)
    print(f"Captured pages: {captured_pages}")
    print(f"Rows written: {len(all_rows)}")
    print(f"Excel file: {OUTPUT_XLSX.resolve()}")
    print(f"Video file: {OUTPUT_VIDEO.resolve()}")


if __name__ == "__main__":
    main()

File: scripts/scrape-tripadvisor-attractions.mjs
/**
 * Opens TripAdvisor China attractions listing in a browser, walks 10 pages
 * via the bottom "Next page" control, extracts attraction id + name from
 * each list card, writes attractionsForAgent.xlsx, and saves a screen recording.
 */
import { chromium } from "playwright";
import ExcelJS from "exceljs";
import { existsSync } from "node:fs";
import { mkdir, readdir } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { execFileSync } from "node:child_process";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, "..");
const START_URL =
  "https://www.tripadvisor.cn/Attractions-g294211-Activities-China.html";
const MAX_PAGES = 10;
const ARTIFACTS_DIR = path.join(REPO_ROOT, "artifacts");
const XLSX_PATH = path.join(REPO_ROOT, "attractionsForAgent.xlsx");
const MP4_PATH = path.join(
  REPO_ROOT,
  "tripadvisor-attractions-scrape-recording.mp4",
);

function delay(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function extractRowsFromPage() {
  return Array.from(document.querySelectorAll("article"))
    .map((article) => {
      const link = article.querySelector('a[href*="Attraction_Review"]');
      const href = link?.getAttribute("href") || "";
      const idMatch = href.match(/-d(\d+)-/);
      const id = idMatch ? idMatch[1] : "";
      const name =
        article.querySelector("h2")?.textContent?.trim() ||
        article.querySelector("img[alt]")?.getAttribute("alt")?.trim() ||
        "";
      return id && name ? { id, name } : null;
    })
    .filter((row) => row != null);
}

async function slowScrollFullPage(page) {
  const height = await page.evaluate(() => document.body.scrollHeight);
  for (let y = 0; y <= height; y += 450) {
    await page.evaluate((yy) => window.scrollTo(0, yy), y);
    await delay(120);
  }
  await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
}

async function extractBatchWithRetry(page) {
  await waitForListingCards(page);
  let batch = await page.evaluate(extractRowsFromPage);
  if (batch.length < 30) {
    await slowScrollFullPage(page);
    await delay(1500);
    batch = await page.evaluate(extractRowsFromPage);
  }
  if (batch.length === 29) {
    await page.reload({ waitUntil: "domcontentloaded", timeout: 90000 });
    await waitForListingCards(page);
    await slowScrollFullPage(page);
    await delay(1500);
    batch = await page.evaluate(extractRowsFromPage);
  }
  return batch;
}

async function waitForListingCards(page) {
  await page.waitForFunction(
    () =>
      document.querySelectorAll('article a[href*="Attraction_Review"]')
        .length >= 10,
    { timeout: 45000 },
  );
  await page
    .locator("article")
    .first()
    .waitFor({ state: "visible", timeout: 15000 });
}

async function main() {
  await mkdir(ARTIFACTS_DIR, { recursive: true });

  const browser = await chromium.launch({
    headless: true,
    args: ["--no-sandbox", "--disable-dev-shm-usage"],
  });

  const context = await browser.newContext({
    viewport: { width: 1280, height: 800 },
    locale: "zh-CN",
    recordVideo: { dir: ARTIFACTS_DIR, size: { width: 1280, height: 800 } },
  });

  const page = await context.newPage();
  await page.goto(START_URL, { waitUntil: "domcontentloaded", timeout: 90000 });
  await waitForListingCards(page);

  const allRows = [];

  for (let pageIndex = 0; pageIndex < MAX_PAGES; pageIndex++) {
    await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
    await delay(800);
    const batch = await extractBatchWithRetry(page);

    for (const row of batch) {
      allRows.push(row);
    }

    if (pageIndex < MAX_PAGES - 1) {
      const next = page.locator('a[aria-label="Next page"]').first();
      await next.scrollIntoViewIfNeeded();
      await next.click();
      await page.waitForLoadState("domcontentloaded");
      await delay(1500);
    }
  }

  await context.close();
  await browser.close();

  const workbook = new ExcelJS.Workbook();
  const sheet = workbook.addWorksheet("景点", {
    views: [{ state: "frozen", ySplit: 1 }],
  });
  sheet.columns = [
    { header: "编号", key: "id", width: 14 },
    { header: "景点名称", key: "name", width: 48 },
  ];
  sheet.getRow(1).font = { bold: true };

  for (const row of allRows) {
    sheet.addRow({ id: row.id, name: row.name });
  }

  await workbook.xlsx.writeFile(XLSX_PATH);

  const videoFiles = (await readdir(ARTIFACTS_DIR)).filter((f) =>
    f.endsWith(".webm"),
  );
  const webmPath =
    videoFiles.length === 1
      ? path.join(ARTIFACTS_DIR, videoFiles[0])
      : path.join(ARTIFACTS_DIR, videoFiles.sort().at(-1) || "recording.webm");

  try {
    execFileSync(
      "ffmpeg",
      [
        "-y",
        "-i",
        webmPath,
        "-c:v",
        "libx264",
        "-preset",
        "veryfast",
        "-crf",
        "28",
        "-pix_fmt",
        "yuv420p",
        "-movflags",
        "+faststart",
        MP4_PATH,
      ],
      { stdio: "inherit" },
    );
  } catch {
    console.warn("ffmpeg failed; leaving Playwright .webm in artifacts/");
  }

  console.log(
    JSON.stringify(
      {
        rows: allRows.length,
        xlsx: XLSX_PATH,
        videoWebm: webmPath,
        videoMp4: existsSync(MP4_PATH) ? MP4_PATH : null,
      },
      null,
      2,
    ),
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
